File: src/LooperAnalyzer/Analysis/OptimizationCandidate.cs
﻿using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using System.Diagnostics;
using System.Linq;

namespace LooperAnalyzer.Analysis
{
    internal sealed class OptimizationCandidate
    {
        public InvocationExpressionSyntax Invocation { get; private set; }
        public string ConsumerMethodName { get; private set; }
        public StatementSyntax ContainingStatement { get; private set; }
        public BlockSyntax ContainingBlock { get; private set; }
        public bool IsInvariantOptimization { get; private set; }
        public bool NeedsRefactoring { get; private set; }
        public bool IsMarkedWithOptimizationTrivia { get; private set; }

        private OptimizationCandidate() { }

        public static OptimizationCandidate FromInvocation(InvocationExpressionSyntax node)
        {
            var memberExpr = node?.Expression as MemberAccessExpressionSyntax;

            Debug.Assert(memberExpr != null);

            var consumerMethod = memberExpr.Name.ToString();

            var parentExpr = node
                .Ancestors()
                .TakeWhile(n => !(n is BlockSyntax))
                .FirstOrDefault(n => n is ExpressionSyntax);
            var parentStmt = node.GetParentStatement();

            var block = node.FirstAncestorOrSelf<BlockSyntax>();

            return new OptimizationCandidate
            {
                Invocation = node,
                ConsumerMethodName = consumerMethod,
                ContainingStatement = parentStmt,
                ContainingBlock = block,
                IsInvariantOptimization = parentExpr == null,
                NeedsRefactoring = !(parentStmt is LocalDeclarationStatementSyntax), // TODO
                IsMarkedWithOptimizationTrivia = parentStmt?.IsMarkedWithOptimizationTrivia() == true
            };
        }
    }
}

File: src/Looper.Core/Analysis/OptimizationCandidate.fs
namespace Looper.Core

open Microsoft.CodeAnalysis
open Microsoft.CodeAnalysis.CSharp.Syntax

type OptimizationCandidate = 
    { Invocation                     : InvocationExpressionSyntax
      ConsumerMethodName             : string
      ContainingStatement            : StatementSyntax option
      ContainingBlock                : BlockSyntax option
      IsInvariantOptimization        : bool
      NeedsRefactoring               : bool
      IsMarkedWithOptimizationTrivia : bool
      SemanticModel                  : SemanticModel }

    static member FromInvocation(model : SemanticModel, node : InvocationExpressionSyntax) = 
        let memberExpr = 
            match node.Expression with
            | :? MemberAccessExpressionSyntax as m -> m
            | _ -> invalidArg "node" "Expected member access"
        
        let consumerMethod = string memberExpr.Name

        let parentExpr = 
            node.Ancestors()
            |> Seq.takeWhile(fun n -> not(n :? BlockSyntax))
            |> Seq.filter(fun n -> n :? ExpressionSyntax)
            |> Seq.tryHead

        let parentStmt = node.GetParentStatement();

        let block = Some <| node.FirstAncestorOrSelf<BlockSyntax>()

        let isMarked = parentStmt |> Option.exists (fun s -> s.IsMarkedWithOptimizationTrivia) 
        let needsRefactoring = parentStmt |> Option.exists(fun s -> not(s :? LocalDeclarationStatementSyntax)) // TODO

        { Invocation = node
          ConsumerMethodName = consumerMethod
          ContainingStatement = parentStmt
          IsInvariantOptimization = parentExpr.IsNone
          ContainingBlock = block
          NeedsRefactoring = needsRefactoring
          IsMarkedWithOptimizationTrivia = isMarked
          SemanticModel = model
        }